### PR TITLE
Add AWS auth method for Vault RA mode

### PR DESCRIPTION
**Pull request by leonweecs**
Implements #1946

#### Name of feature:

Add AWS auth method for Vault RA mode.

The available `authOptions` maps directly to the Vault AWS auth [package](https://github.com/hashicorp/vault/blob/main/api/auth/aws/aws.go).

#### Pain or issue this feature alleviates:

Please refer to #1946.

#### Why is this important to the project (if not answered above):

N/A

#### Is there documentation on how to use this feature? If so, where?

Yes, a separate PR will be raised for the docs repo.

https://github.com/smallstep/docs/pull/346

#### In what environments or workflows is this feature supported?

- AWS
- Vault RA mode

#### In what environments or workflows is this feature explicitly NOT supported (if any)?

N/A

#### Supporting links/other PRs/issues:

- #1946
- https://developer.hashicorp.com/vault/docs/auth/aws

💔Thank you!

### CODE REVIEW

File: cas/vaultcas/auth/aws/aws.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		default:
+			return nil, fmt.Errorf("unknown SignatureType type: %s, valid options are 'pkcs7', 'identity' and 'rsa2048'", opts.SignatureType)
+		}
+
+		if opts.Nonce != "" {
+			loginOptions = append(loginOptions, aws.WithNonce(opts.Nonce))
+		}
+	default:
+		return nil, fmt.Errorf("unknown awsAuthType: %s, valid options are 'iam' and 'ec2'", opts.AwsAuthType)
+	}
```
Comment (hslatman) [inline cas/vaultcas/auth/aws/aws.go:76]:
```suggestion
		return nil, fmt.Errorf("unknown awsAuthType %q; valid options are 'iam' and 'ec2'", opts.AwsAuthType)
	}
```

File: cas/vaultcas/auth/aws/aws.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+			loginOptions = append(loginOptions, aws.WithPKCS7Signature())
+		case "identity":
+			loginOptions = append(loginOptions, aws.WithIdentitySignature())
+		case "rsa2048":
+			loginOptions = append(loginOptions, aws.WithRSA2048Signature())
+		case "":
+			// no-op
+		default:
+			return nil, fmt.Errorf("unknown SignatureType type: %s, valid options are 'pkcs7', 'identity' and 'rsa2048'", opts.SignatureType)
+		}
```
Comment (hslatman) [inline cas/vaultcas/auth/aws/aws.go:69]:
```suggestion
			return nil, fmt.Errorf("unknown SignatureType type %q; valid options are 'pkcs7', 'identity' and 'rsa2048'", opts.SignatureType)
		}
```

File: cas/vaultcas/auth/aws/aws_test.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+	return u, client
+}
+
+func TestAws_LoginMountPaths(t *testing.T) {
+	caURL, _ := testCAHelper(t)
+
+	config := vault.DefaultConfig()
+	config.Address = caURL.String()
+	client, _ := vault.NewClient(config)
```
Comment (hslatman) [inline cas/vaultcas/auth/aws/aws_test.go:65]: Is it necessary to create the `client` in both parts? It seems the second time it uses the same inputs, so it looks like the returned `client` would be OK to use too?

Comment (hslatman) [reply cas/vaultcas/auth/aws/aws_test.go:65]: Looks like it's the same for the other tests. Anyway, seems like the second time is not needed here.

**Comment by hslatman** [comment]
Don't forget to fix the conflicts by rebasing or some other way

File: cas/vaultcas/auth/aws/aws_test.go (hunk 3 above)
Comment (leonweecs) [reply cas/vaultcas/auth/aws/aws_test.go]: Yeah agree that it is redundant to create another `client`, we can use the one from the helper directly. I have updated accordingly.

**Review by hslatman — APPROVED** [review]

**Pull request merged**